File: Tuning/sample_code/02_timer-res/src/cpp/main.cpp
// Copyright 2024 Research Organization for Information Science and Technology 
//----------------------------------------------------------------------
//  Check resolution of your timer 
//  Author:      Yukihiro Ota (yota@rist.or.jp)
//  Description: The original idea comes from Wadleigh and Crawford,
//               "Software Optimization for HPC: Creating Faster 
//                Applications" (2000) pp.136-138 
//----------------------------------------------------------------------
#include <chrono>
#include <cstdio>
#include <cstdlib>
//#include <iostream>

int func ( const int nn ) ;

//----------------------------------------------------------------------
//  main                                                              
//----------------------------------------------------------------------
int main ( int argc, char* argv[] )
{
  int nn ;
  double t ;

  printf ("--------------------------------------------------------\n");

  // check resolution of wallclock timer with steady_clock
  nn = 0;
  t = 0.0;
  while ( t <= 0.0 ) {
    nn++;
    const auto elp1 = std::chrono::steady_clock::now();
    func (nn);
    const auto elp2 = std::chrono::steady_clock::now();
    const std::chrono::duration<double> elapsed = elp2 - elp1;
    t = static_cast<double>(elapsed.count());
    //std::cout << elapsed.count() << std::endl;
    //std::cout << t << std::endl;
  }

  printf ("[Check resolution of steady_clock]\n") ;
  printf ("It took %7d iterations to generate a none-zero time\n", nn) ;
  if ( nn == 1 ) {
    printf (" timer resolution less than or equal to %15.9f\n", t) ;      
  } else {
    printf (" timer resolution is %15.9f sec.\n", t) ;      
  }

  printf ("--------------------------------------------------------\n");

  // check resolution of wallclock timer with high_resolution_clock
  nn = 0;
  t = 0.0;
  while ( t <= 0.0 ) {
    nn++;
    const auto elp1 = std::chrono::high_resolution_clock::now();
    func (nn);
    const auto elp2 = std::chrono::high_resolution_clock::now();
    const std::chrono::duration<double> elapsed = elp2 - elp1;
    t = static_cast<double>( elapsed.count() );
  }

  printf ("[Check resolution of high_resolution_clock]\n") ;
  printf ("It took %7d iterations to generate a none-zero time\n", nn) ;
  if ( nn == 1 ) {
    printf (" timer resolution less than or equal to %15.9f\n", t) ;      
  } else {
    printf (" timer resolution is %15.9f sec.\n", t) ;      
  }

  printf ("--------------------------------------------------------\n");

  /* finalization */
  return EXIT_SUCCESS ;
}
/*--------------------------------------------------------------------*/
/*  func                                                              */
/*--------------------------------------------------------------------*/
int func ( const int nn ) 
{
  int i, j ;
  i = 0 ;
  for ( j = 0 ; j < nn ; ++j ) i++ ;
  return i ;
}

File: Tuning/sample_code/02_timer-res/src/fortran/main.f90
! Copyright 2024 Research Organization for Information Science and Technology
!*************************************************************************
!  Check resolution of your timer
!  Author:      Yukihiro Ota (yota@rist.or.jp)
!  Description: The original idea comes from Wadleigh and Crawford,
!               "Software Optimization for HPC: Creating Faster 
!                Applications" (2000) pp.136-138 
!*************************************************************************

!=========================================================================
!   module: mytype 
!=========================================================================
module mytype
  implicit none
  integer,parameter :: SP = kind(1.0)
  integer,parameter :: DP = selected_real_kind(2*precision(1.0_SP))
end module mytype
!=========================================================================
!   main 
!=========================================================================
program main

  use mytype,only: DP

  implicit none

  !! local var
  integer :: i, nn, cnt1, cnt2, cnt_rate, cnt_max

  real(kind=DP) :: tval, tval0 
  real(kind=DP),parameter :: ZERO = 0.0_DP

  !! interface to function
  interface
    
    function func(nn)
      integer :: func
      integer,intent(in) :: nn
    end function func

  end interface

  write(6,'("----------------------------------------------------------")')

  !! check resolution of wallclock timer  
  tval = -1.0_DP 
  nn = 0
  do while ( tval .le. ZERO )
    nn = nn + 1
    call system_clock (cnt1)
    i = func(nn)
    call system_clock (cnt2, cnt_rate, cnt_max)
    if ( cnt2 .ge. cnt1 ) then
       tval = (cnt2 - cnt1) / dble(cnt_rate)
    else
       tval = (cnt2 - cnt1 + cnt_max + 1) / dble(cnt_rate)
    end if
  end do

  write(6,'("[Check resolution of wallclock timer]")')
  write(6,'("It took ",1I7," &
&   iterations to generate a none-zero time")') nn 
  if ( nn == 1 ) then
    write(6,'(" timer resolution less than or equal to ",1F15.9)') tval
  else
    write(6,'(" timer resolution is ",1F15.9," sec.")') tval
  end if

  write(6,'("----------------------------------------------------------")')

  !! check resolution of cpu timer  
  tval = -1.0_DP 
  nn = 0
  do while ( tval .le. ZERO )
    nn = nn + 1
    call cpu_time (tval0)
    i = func(nn)
    call cpu_time (tval)
    tval = tval - tval0
  end do

  write(6,'("[Check resolution of cpu timer]")')
  write(6,'("It took ",1I7," &
&   iterations to generate a none-zero time")') nn 
  if ( nn == 1 ) then
    write(6,'(" timer resolution less than or equal to ",1F15.9)') tval
  else
    write(6,'(" timer resolution is ",1F15.9," sec.")') tval
  end if

  write(6,'("----------------------------------------------------------")')

  stop
end program main
!==========================================================================
!   func 
!==========================================================================
function func (nn)

  implicit none

  integer :: func
  integer,intent(in) :: nn

  integer :: j

  func = 0
  do j = 1, nn
    func = func + 1
  end do
end function func
